atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/HostnamesByIpSupplier.java

```diff
@@ -0,0 +1,99 @@
+
+import org.apache.cassandra.thrift.Compression;
+import org.apache.cassandra.thrift.ConsistencyLevel;
+import org.apache.cassandra.thrift.CqlRow;
+import org.apache.cassandra.thrift.KsDef;
+import org.apache.cassandra.thrift.NotFoundException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Iterables;
+import com.palantir.atlasdb.encoding.PtBytes;
+import com.palantir.atlasdb.keyvalue.cassandra.CassandraClient;
+import com.palantir.atlasdb.keyvalue.cassandra.CqlQuery;
+import com.palantir.common.base.FunctionCheckedException;
+import com.palantir.common.pooling.PoolingContainer;
+import com.palantir.logsafe.SafeArg;
+
+public final class HostnamesByIpSupplier implements Supplier<Map<String, String>> {
+    private static final Logger logger = LoggerFactory.getLogger(HostnamesByIpSupplier.class);
+
+    private static final String SYSTEM_PALANTIR_KEYSPACE = "system_palantir";
+    private static final String HOSTNAMES_BY_IP_TABLE = "hostnames_by_ip";
+    private static final String HOSTNAME_COLUMN = "hostname";
+    private static final String IP_COLUMN = "ip";
+
+    private final Supplier<PoolingContainer<CassandraClient>> randomGoodHostSupplier;
+
+    public HostnamesByIpSupplier(Supplier<PoolingContainer<CassandraClient>> randomGoodHostSupplier) {
+        this.randomGoodHostSupplier = randomGoodHostSupplier;
+    }
+
+    @Override
+    public Map<String, String> get() {
+        try {
+            return randomGoodHostSupplier.get().runWithPooledResource(getHostnamesByIp());
+        } catch (Exception e) {
+            logger.warn("Could not get hostnames by ip from Cassandra", e);
+            return ImmutableMap.of();
+        }
+    }
+
+    public FunctionCheckedException<CassandraClient, Map<String, String>, Exception> getHostnamesByIp() {
+        return client -> {
+            KsDef systemPalantir;
+            try {
+                systemPalantir = client.describe_keyspace(SYSTEM_PALANTIR_KEYSPACE);
+            }  catch (NotFoundException e) {
+                logger.debug("Did not find keyspace with hostnames by ip, moving on without them");
+                return ImmutableMap.of();
+            }
+            if (systemPalantir.getCf_defs().stream().noneMatch(cfDef -> cfDef.name.equals(HOSTNAMES_BY_IP_TABLE))) {
+                logger.debug("Did not find table with hostnames by ip, moving on without them");
+                return ImmutableMap.of();
+            }
+
+            CqlQuery query = CqlQuery.builder()
+                    .safeQueryFormat("SELECT * FROM \"%s\".\"%s\";")
+                    .addArgs(
+                            SafeArg.of("keyspace", SYSTEM_PALANTIR_KEYSPACE),
+                            SafeArg.of("table", HOSTNAMES_BY_IP_TABLE))
+                    .build();
+
+            return client.execute_cql3_query(query, Compression.NONE, ConsistencyLevel.LOCAL_ONE).getRows().stream()
```

Comment:
It is populated out of band internally and yea, we really don't need any real guarantees on this table, even if the data is out of date / wrong, it won't matter too much, since we read the real ring from Cassandra anyways.